File: src/main/scala/actors/MowerActor.scala
package actors

import akka.actor.{ActorRef, Actor, ActorLogging, Props}
import model.{Command, Forward, Mower}

class MowerActor(parent: ActorRef) extends Actor with ActorLogging {

  def receive = {

    // TODO 1.2 add four case
    // TODO 1.2.1 first one has to handle ExecuteCommands message

    log.debug(s"Executing instructions for <$mower>, remaining: <$commands>")

    commands match {
      case Nil => // TODO 1.2.1.1 no commands left, send AllCommandsExecutedOn message to the parent actor
      case _ => // TODO 1.2.1.2 call handleRemainingCommands function
    }

    // TODO 1.2.2 second one has to handle PositionAllowed message and send to itself ExecuteCommands message
    log.debug(s"Position <${newState.pos}> authorized, remaining:<${commands.tail}>")

    // TODO 1.2.3 third one has to handle PositionRejected message and send to itself ExecuteCommands message

    // TODO 1.2.4 fourth one has to handle TerminateProcessing message and stop the current actor
    log.debug(s"Terminating $mower")

  }

  def handleRemainingCommands(mower: Mower, commands: List[Command], retry: Int) = {
    commands match {
      case command@Forward :: tail =>
        log.debug(s"Going forward on <$mower>, remaining:<$tail>, retry:<$retry>")
        val newState = mower.forward
        parent ! MowerMessages.RequestAuthorisation(mower, newState, commands, retry)

      case command :: tail =>
        log.debug(s"Rotating:<$command>, remaining:<$tail>")
        val newState = mower.rotate(command)
        self ! MowerMessages.ExecuteCommands(newState, commands.tail, 0)
    }
  }
}

object MowerMessages {

  case class ExecuteCommands(mower: Mower, commands: List[Command], retry: Int)

  case class RequestAuthorisation(currentState: Mower, newState: Mower, commands: List[Command], retry: Int)

  case class PositionAllowed(mower: Mower, commands: List[Command])

  case class PositionRejected(mower: Mower, commands: List[Command], retry: Int)

  case object PrintPosition

  case class TerminateProcessing(mower: Mower)

  case class AllCommandsExecutedOn(mower: Mower)

}

object MowerActor {

  // TODO 1.1 finish to implement the props function
  // def props(parent: ActorRef) =

}

File: src/main/java/fr/xebia/mower/MowerActor.java
package fr.xebia.mower;

import akka.actor.AbstractLoggingActor;
import akka.actor.ActorRef;
import akka.actor.Props;
import akka.japi.pf.ReceiveBuilder;

import static java.lang.String.format;

public class MowerActor extends AbstractLoggingActor {

    private ActorRef parent;

    public MowerActor(ActorRef parent) {
        this.parent = parent;
        receive(
                ReceiveBuilder
                        .match(MowerMessages.ExecuteCommands.class, this::onExecuteCommands)
                        .match(MowerMessages.PositionAllowed.class, this::onPositionAllowed)
                        .match(MowerMessages.PositionRejected.class, this::onPositionRejected)
                        .match(MowerMessages.TerminateProcessing.class, this::onTerminateProcessing)
                        .build()
        );
    }

    public static Props props(ActorRef parent) {
        return Props.create(MowerActor.class, parent);
    }

    private void onExecuteCommands(MowerMessages.ExecuteCommands message) {
        log().debug(format("Executing instructions for <$mower>, remaining: <%s>", message.mower.commands()));
        if (message.mower.commands().isEmpty()) {
            handleNoMoreCommands(message);
        } else {
            handleRemainingCommands(message.mower, message.retry);
        }
    }

    private void onPositionAllowed(MowerMessages.PositionAllowed message) {
        final Mower newState = message.mower.dequeueCommand();
        log().debug(format("Position <%s> authorized, remaining:<%s>", message.mower.position(), newState.commands()));
        self().tell(new MowerMessages.ExecuteCommands(newState, 0), self());
    }

    private void onPositionRejected(MowerMessages.PositionRejected message) {
        self().tell(new MowerMessages.ExecuteCommands(message.mower, message.retry), self());
    }

    private void onTerminateProcessing(MowerMessages.TerminateProcessing message) {
        log().debug(format("Terminating %s", message.mower));
        context().stop(self());
    }

    private void handleNoMoreCommands(MowerMessages.ExecuteCommands message) {
        parent.tell(new MowerMessages.AllCommandsExecutedOn(message.mower), self());
    }

    private void handleRemainingCommands(Mower mower, int retry) {
        log().debug(format("Going forward on <%s>, remaining:<%s>, retry:<%d>", mower, mower.commands().tail(), retry));
        final Mower newState = mower.dequeueCommand();
        if (mower.commandNeedToAskAuth()) {
            parent.tell(new MowerMessages.RequestAuthorisation(mower, newState, retry), self());
        } else {
            self().tell(new MowerMessages.ExecuteCommands(newState, 0), self());
        }
    }
}
